crates/supervisor/core/src/safety_checker/task.rs

```diff
@@ -0,0 +1,255 @@
+    ///
+    /// On each iteration:
+    /// - Tries to promote the next eligible block
+    /// - Waits for `interval` if promotion fails
+    /// - Exits when [`cancel_token`](CancellationToken) is triggered
+    pub async fn run(self) {
+        info!(
+            target: "safety_checker",
+            chain_id = self.chain_id,
+            target_level = %self.target_level,
+            "Started safety checker");
+
+        let checker = CrossSafetyChecker::new(&*self.provider);
+
+        loop {
+            tokio::select! {
+                _ = self.cancel_token.cancelled() => {
+                    info!(target: "safety_checker", chain_id = self.chain_id,target_level = %self.target_level, "Canceled safety checker");
+                    break;
+                }
+
+                _ = async {
+                    match self.promote_next_block(&checker) {
+                        Ok(block_info) => {
+                            debug!(
+                                target: "safety_checker",
+                                chain_id = self.chain_id,
+                                target_level = %self.target_level,
+                                %block_info,
+                                "Promoted next candidate block"
+                            );
+                        }
+                        Err(err) => {
+                            warn!(
+                                target: "safety_checker",
+                                chain_id = self.chain_id,
+                                target_level = %self.target_level,
+                                %err,
+                                "Error promoting next candidate block"
+                            );
+                            tokio::time::sleep(self.interval).await;
+                        }
+                    }
+                } => {}
+            }
+        }
+
+        info!(target: "safety_checker", chain_id = self.chain_id, target_level = %self.target_level, "Stopped safety checker");
+    }
+
+    // Attempts to promote the next block at the current safety level,
+    // after validating cross-chain dependencies.
+    fn promote_next_block(
+        &self,
+        checker: &CrossSafetyChecker<'_, P>,
+    ) -> Result<BlockInfo, CrossSafetyError> {
+        let candidate = self.find_next_promotable_block()?;
+
+        checker.verify_block_dependencies(self.chain_id, candidate, self.target_level)?;
+
+        // TODO: Add more checks in future
+
+        self.provider.update_safety_head_ref(self.chain_id, self.target_level, &candidate)?;
+
+        Ok(candidate)
+    }
+
+    // Finds the next block that is eligible for promotion at the current target level.
+    fn find_next_promotable_block(&self) -> Result<BlockInfo, CrossSafetyError> {
+        let promotion_boundary = self.promotion_upper_bound()?;
+
+        let current_head = self.provider.get_safety_head_ref(self.chain_id, self.target_level)?;
```

Comment:
will any other threads be promoting blocks for the same chain in parallel? otherwise we can optimise this by storing the current block number in a field of this type, and updating it when a block is promoted. that way we only need one db/cache read (upper head) each time this thread is polled. would this optimisation be reorg safe?

Reply:
I did it this way only considering the re-org. I think we surely will revisit this again while implementing re-org and will try to optimize. That's why I tried to avoid maintaining the current state for the block number.